managers/backup-manager/BackupWatcher.js

```diff
@@ -0,0 +1,151 @@
+        resourceId: changedOptions.instance_guid,
+        operationName: CONST.APISERVER.ANNOTATION_NAMES.BACKUP,
+        operationType: CONST.APISERVER.ANNOTATION_TYPES.BACKUP,
+        operationId: changedOptions.guid
+      })
+      .then(options => {
+        return Promise.try(() => {
+          const plan = catalog.getPlan(options.plan_id);
+          return bm.createManager(plan);
+        }).then(manager => manager.deleteBackup(options));
+      });
+  }
+
+  /**
+   * @description Registers backup watcher with worker callback
+   */
+  registerWatcher() {
+    logger.info(`Registering Backup watcher`);
+    const queryString = `state in (${CONST.APISERVER.RESOURCE_STATE.IN_QUEUE},${CONST.OPERATION.ABORT},${CONST.APISERVER.RESOURCE_STATE.DELETE})`;
+    return eventmesh.apiServerClient.registerWatcher('backup', 'defaultbackup', this.worker, queryString)
+      .then(stream => {
+        logger.info(`Successfully set watcher with query string:`, queryString);
+        return setTimeout(() => {
+          try {
+            logger.info(`Refreshing stream after ${CONST.APISERVER.WATCHER_REFRESH_INTERVAL}`);
+            stream.abort();
+            return this.registerWatcher();
+          } catch (err) {
+            logger.error('Error caught in the set timout callback');
+            throw err;
+          }
+        }, CONST.APISERVER.WATCHER_REFRESH_INTERVAL);
+      })
+      .catch(e => {
+        logger.error('Failed registering watcher with error:', e);
+        throw e;
+      });
+  }
+  /**
+   * @description This method does following in order
+   * 1. Tries to acquire processing lock => In case of HA, Only one backup manager process will process the request
+   * 2. Processes the request based on the state of the changed resource
+   * 3. Release processing lock
+   * @param {object} change - Change object that comes as part of apiserver watch event
+   */
+
+  worker(change) {
+
+    logger.info('Changed Resource:', change);
+    logger.debug('Changed resource options:', change.object.spec.options);
+    const changeObjectBody = change.object;
+    const changedOptions = JSON.parse(changeObjectBody.spec.options);
+    logger.debug('Changed resource options(parsed):', changedOptions);
+    let processingLockConflict = false;
+    if (changeObjectBody.metadata.annotations && changeObjectBody.metadata.annotations.lockedByManager === config.broker_ip) {
+      // Don't have to process as this is event generated by acquireProcessLock
+      logger.debug(`Backup request for guid ${changedOptions.guid} is already being processed`);
+    } else {
+      // Acquire processing lock so that in HA scenerio, only one backup-manager process processes the request
+      return Promise.try(() => {
+          if (!changeObjectBody.metadata.annotations || changeObjectBody.metadata.annotations.lockedByManager === '') {
+            return BaseManager.acquireProcessingLock(changeObjectBody)
```

Comment:
shouldnt this acquiring lock part move to BaseManager?

Reply:
I would recommend the following. You can have the following method in BaseManager:
```
handleResource(change){
return this._preProcessResource(change)
            .then(()=> this.processResource(change))
            .then(()=> this.postProcessResource(change))
             .catch((err)=> logger.error(`Error occurred ....`)); //Dont need to care what type of error just log.
}

_preProcessResource(){
//Code about acquiring processing lock goes here.
}

processResource(change){
  throw 'Must be implemented by subclass'; // Individual base manage managers just implement this.
}
_postProcessResource(){
 //code to handle release processing lock  goes here. 
}
```
Now code is really clean and individual managers just implement 'processResource' and register 'handleResource' which is part of BaseManager.